File: src/test/java/local/hillclimbing/HillClimbingTest.java
package local.hillclimbing;

import core.Heuristic;
import local.queens.NumberOfClashesHeuristic;
import local.queens.State;
import org.junit.Assert;
import org.junit.Ignore;
import org.junit.Test;

public class HillClimbingTest {
	@Test
	public void randomRestart() {
		Heuristic<State> heuristic = new NumberOfClashesHeuristic();
		RandomRestart<State> search = new RandomRestart<>(heuristic, null, 0, 0);
		search.run(() -> new State(8));
		
		// Test the functionality for limiting the number of iterations.
		final int maximumNrIterations = 2;
		for (int testIteration = 0; testIteration < 10; testIteration++) {
			RandomRestart<State> searchWithMaxIterations = new RandomRestart<>(heuristic, maximumNrIterations, 0, 0);
			searchWithMaxIterations.run(() -> new State(8));
			Assert.assertTrue(searchWithMaxIterations.nrIterationsPerformed <= maximumNrIterations);
		}
		
		// Test the functionality for quality margins.
		final double qualityMargin = 1;
		for (int testIteration = 0; testIteration < 10; testIteration++) {
			RandomRestart<State> searchWithMaxIterations = new RandomRestart<>(heuristic, null, qualityMargin, 0);
			State state = searchWithMaxIterations.run(() -> new State(4));
			Assert.assertTrue(heuristic.determineQualityScore(state) <= heuristic.getBestPossibleScore() + qualityMargin + .0001);
		}
	}
	
	@Test
	public void steepestAscent() {
		State state = new State(8);
		Heuristic<State> heuristic = new NumberOfClashesHeuristic();
		SteepestAscent<State> climb = new SteepestAscent<>(state, heuristic, 100);
		climb.run();
	}
	
	@Test
	public void stochastic() {
		State state = new State(8);
		Heuristic<State> heuristic = new NumberOfClashesHeuristic();
		Stochastic<State> climb = new Stochastic<>(state, heuristic, 100);
		climb.run();
	}
	
	@Test
	public void simulatedAnnealing() {
		State state = new State(8);
		Heuristic<State> heuristic = new NumberOfClashesHeuristic();
		SimulatedAnnealing<State> climb = new SimulatedAnnealing<>(state, heuristic, .05);
		climb.run();
	}
	
	@Ignore
	@Test
	public void big() {
		Heuristic<State> heuristic = new NumberOfClashesHeuristic();
		RandomRestart<State> searchWithMaxIterations = new RandomRestart<>(heuristic, 100, 0, 20);
		searchWithMaxIterations.run(() -> new State(88));
	}
}

File: src/test/scala/nl/calmamity/search/local/hillclimbing/HillClimbingTest.scala
package nl.calmamity.search.local.hillclimbing

import nl.calmamity.search.SearchTest
import nl.calmamity.search.local.queens.{NumberOfClashesHeuristic, State}
import nl.calmamity.search.local.supereffective.{EffectivenessHeuristic, Team}

class HillClimbingTest extends SearchTest {
	"Random restart" should "climb a couple of hills" in {
		val heuristic = new NumberOfClashesHeuristic()
		val search = new RandomRestart[State](heuristic, 0, None, 0)
		search.run(() => State(8))
		
		// Test the functionality for limiting the number of iterations.
//		val maximumNrIterations = 2
//		for (testIteration <- 1 to 10) {
//			val searchWithMaxIterations = new RandomRestart[State](heuristic, 0, Some(maximumNrIterations), 0)
//			searchWithMaxIterations.run(() => State.apply(8))
//			Assert.assertTrue(searchWithMaxIterations.numIterationsPerformed <= maximumNrIterations)
//		}
		
		// Test the functionality for quality margins.
		val qualityMargin = 1
		for (_ <- 1 to 10) {
			val searchWithMaxIterations = new RandomRestart[State](heuristic, 0, None, qualityMargin)
			val state = searchWithMaxIterations.run(() => State(4))
			assert(heuristic.determineIsOptimalScore(heuristic.determineQualityScore(state), qualityMargin))
		}
	}
	
	it should "do something big" ignore {
		val heuristic = new NumberOfClashesHeuristic()
		val searchWithMaxIterations = new RandomRestart[State](heuristic, 20, Some(100), 0)
		searchWithMaxIterations.run(() => State(88))
	}
	
	it should "do something super-effective" in {
		val randomRestart = new RandomRestart[Team](new EffectivenessHeuristic(), 10, Some(50), 0)
		randomRestart.run(() => Team(2))
	}
	
	"Steepest ascent" should "climb a single hill" in {
		val state = State(8)
		val heuristic = new NumberOfClashesHeuristic()
		val climb = new SteepestAscent[State](state, heuristic, 100)
		climb.run()
	}
	
	it should "find the superest effective" in {
		val team = Team(6)
		val heuristic = new EffectivenessHeuristic()
		val climb = new SteepestAscent[Team](team, heuristic, 10)
		climb.run()
	}
	
	"Stochastic" should "maybe climb a hill" in {
		val state = State(8)
		val heuristic = new NumberOfClashesHeuristic()
		val climb = new Stochastic[State](state, heuristic, 100)
		climb.run()
	}
	
	"Simulated annealing" should "do something like annealing" in {
		val state = State(8)
		val heuristic = new NumberOfClashesHeuristic()
		val climb = new SimulatedAnnealing[State](state, heuristic, .05, 100)
		climb.run()
	}
	
	it should "do something big" ignore {
		val state = State(888)
		val heuristic = new NumberOfClashesHeuristic()
		val climb = new SimulatedAnnealing[State](state, heuristic, 1d / 5000, 1000)
		climb.run()
	}
	
	it should "find the superest effective" in {
		val heuristic = new EffectivenessHeuristic()
		val endStates = for (iteration <- 1 to 100) yield {
			log.info(s"Starting iteration #$iteration.")
			val team = Team(3)
			val climb = new SimulatedAnnealing[Team](team, heuristic, 1d / 5000, 1000)
			climb.run()
		}
		
		val endStatesWithScores = endStates
			.distinct
			.map {
				team =>
					(team, heuristic.determineQualityScore(team))
			}
			.sortWith {
				case ((_, score1), (_, score2)) =>
					score1 > score2
			}
			
		log.info(s"Best scoring teams:\n${endStatesWithScores.mkString("\n")}")
	}
}
